h2quic/response_writer.go

```diff
@@ -86,12 +98,129 @@ func (w *responseWriter) Flush() {}
+		}
+		u.Scheme = wantScheme
+		u.Host = w.requestHost
+	} else {
+		if u.Scheme != wantScheme {
+			return fmt.Errorf("cannot push URL with scheme %q from request with scheme %q", u.Scheme, wantScheme)
+		}
+		if u.Host == "" {
+			return errors.New("URL must have a host")
+		}
+	}
+	for k := range opts.Header {
+		if strings.HasPrefix(k, ":") {
+			return fmt.Errorf("promised request headers cannot include pseudo header %q", k)
+		}
+		switch strings.ToLower(k) {
+		case "content-length", "content-encoding", "trailer", "te", "expect", "host":
+			return fmt.Errorf("promised request headers cannot include %q", k)
+		}
+	}
+	if err = checkValidHTTP2RequestHeaders(opts.Header); err != nil {
+		return err
+	}
+	authority := u.Host
+	path := u.RequestURI()
+	contentLengthStr := "0" // a request does not have content
+
+	// Construct HTTP headers for request in push promise
+	var headers bytes.Buffer
+	enc := hpack.NewEncoder(&headers)
+	enc.WriteField(hpack.HeaderField{Name: ":method", Value: opts.Method})
+	enc.WriteField(hpack.HeaderField{Name: ":path", Value: path})
+	enc.WriteField(hpack.HeaderField{Name: ":authority", Value: authority})
+	enc.WriteField(hpack.HeaderField{Name: ":scheme", Value: u.Scheme})
+	for k, v := range opts.Header {
+		for index := range v {
+			enc.WriteField(hpack.HeaderField{Name: strings.ToLower(k), Value: v[index]})
+		}
+	}
+
+	// Get new data stream
+	newDataStream, err := w.session.OpenStream()
+	if err != nil {
+		if err == qerr.TooManyOpenStreams {
+			return http2.ErrPushLimitReached
+		}
+		return err
+	}
+	newDataStreamID := newDataStream.StreamID()
+	if newDataStreamID%2 != 0 {
+		return fmt.Errorf("Cannot push on client side stream")
+	}
+
+	// Write push promise header
+	utils.Debugf("Sending PUSH_PROMISE for target '%s', promised on stream %d", target, newDataStreamID)
+	headerFramer := http2.NewFramer(w.headerStream, nil)
+	w.headerStreamMutex.Lock()
+	err = headerFramer.WritePushPromise(http2.PushPromiseParam{
+		StreamID:      uint32(w.dataStreamID),
+		PromiseID:     uint32(newDataStreamID),
+		BlockFragment: headers.Bytes(),
+		EndHeaders:    true,
+	})
+	w.headerStreamMutex.Unlock() // Do not defer as we will first call ServeHTTP(), which will need the mutex
+	if err != nil {
+		return err
+	}
+
+	// golang net/http2 constructs a 'fake' http2 request and feeds it to the serve() loop to let it be processed as a normal request.
+	// But we will, for now, just serveHTTP() it here:
+	pushRequest, err := requestFromHTTPHeader(opts.Header, path, authority, opts.Method, contentLengthStr)
+	if err != nil {
+		return err
+	}
+	pushRequest.RemoteAddr = w.session.RemoteAddr().String()
+	reqBody := newRequestBody(newDataStream)
+	pushRequest.Body = reqBody
+	pushRequestResponseWriter := newResponseWriter(w.headerStream, w.headerStreamMutex, newDataStream, newDataStreamID, w.session, w.handlerFunc, w.requestHost)
+	serveHTTP(w.handlerFunc, pushRequestResponseWriter, pushRequest, true, reqBody)
```

Comment:
Are you sure about the `streamEnded == true` here? As you just created the stream above, it's still open from the client side. If the client doesn't send an explicit close it would therefore stay open. Does Chrome do that? If not, we need to ensure it's properly closed (probably using `CloseRemote`), so that we don't leak it.